### PR TITLE
Optimize CI time

### PR DESCRIPTION
**Pull request by reakaleek**
## What does this pull request do?

~Adds the `--cache-from` option.~
~This way the build time will be shorter. If there are changes in the Dockerfile, it will still be built utilising the cache.~
~Will only work after https://github.com/elastic/apm-pipeline-library/pull/2174 is merged.~

Use daily built public docker image instead of building on every run.

## Related issues

- depends on https://github.com/elastic/apm-pipeline-library/pull/2174

### CODE REVIEW

**Comment by basepi** [comment]
How will this work for PRs? Is this only for nightly tests?

I think what @beniwohli was requesting in https://github.com/elastic/observability-robots/issues/1623 was that we still create the docker image on each test run, but we do it up front and use the same image for each of the matrix runs (as we currently build it in every matrix run individually). But I don't think https://github.com/elastic/apm-pipeline-library/pull/2174 is creating images for every commit pushed to a PR, so I'm not sure it solves the problem.

But I could be missing something! Thanks for your work on this!

**Comment by beniwohli** [comment]
@basepi these base images don't really change from PR to PR. Updating them daily or even weekly is fine (to pick up changes in the underlying Python base images). We still execute a `pip install -r reqs-base.txt` on every run, so if a PR adds a new test dependency, it'll be installed.

**Comment by reakaleek** [comment]
@basepi 

I'm with you here. This setup is not the most ideal. 

My preferred setup would be utilising the [--cache-from](https://docs.docker.com/engine/reference/commandline/build/#cache-from) option of the `docker build` command. This way we would benefit from caching the image and the ability to rebuild the image on every commit if there are changes.

However, we have this limitation that we cannot use secrets in workflows that are triggered on forked PRs.
Hence, we cannot do a docker login and won't be able to push the image.

As @beniwohli mentioned the base image does not change that much. But this setup will surely create discrepancy when you want to test changes on the docker image itself.

Let me experiment a bit more and I will try to figure out if I can somehow GitHub's artifacts API to achieve the desired caching behavior.

**Comment by reakaleek** [comment]
@basepi and @beniwohli 

I think I got something intereseting here in https://github.com/elastic/apm-agent-python/pull/1793 which is inspired by https://depot.dev/blog/docker-layer-caching-in-github-actions

In this [run](https://github.com/elastic/apm-agent-python/actions/runs/4615344598/jobs/8159136799) you can see the building utilizes the docker layer cache and its time is reduced to ~16 seconds.

cc @elastic/observablt-ci

File: tests/scripts/docker/run_tests.sh [hunk 1]
```diff
@@ -47,7 +47,7 @@ fi
 
 # CASS_DRIVER_NO_EXTENSIONS is set so we don't build the Cassandra C-extensions,
 # as this can take several minutes
-docker build --build-arg PYTHON_IMAGE=${1/-/:} -t apm-agent-python:${1} . # replace - with : to get the correct docker image
+
```
Comment (v1v) [inline tests/scripts/docker/run_tests.sh:50]: Given a PR that changes the `Dockerfile`, how will it be possible to change the PR in the CI? I'm not sure whether that's something we do often.

Comment (reakaleek) [reply tests/scripts/docker/run_tests.sh]: I just added this a couple of minutes ago.

https://github.com/elastic/apm-agent-python/pull/1792/commits/1cd662e614829d9062c7e560b47eaa02b5864df4

this way we can utilise the cache and still build it on a individual commit.

**Comment by reakaleek** [comment]
Alright, I think I got a good balance now.

We can build the images daily and publish it dockerhub like suggested in https://github.com/elastic/apm-pipeline-library/pull/2174.

Here, we can utilise the `--cache-from` option, which will speed up the build significantly, but you will still be able to make changes to the docker image for individual commits, if necessary.

File: tests/scripts/docker/run_tests.sh (hunk 1 above)
Comment (reakaleek) [reply tests/scripts/docker/run_tests.sh]: https://github.com/elastic/apm-agent-python/pull/1792#issuecomment-1496992110

File: tests/scripts/docker/run_tests.sh [hunk 2]
```diff
@@ -47,7 +47,7 @@ fi
 
 # CASS_DRIVER_NO_EXTENSIONS is set so we don't build the Cassandra C-extensions,
 # as this can take several minutes
-docker build --build-arg PYTHON_IMAGE=${1/-/:} -t apm-agent-python:${1} . # replace - with : to get the correct docker image
+DOCKER_BUILDKIT=1 docker build --cache-from=elasticobservability/apm-agent-python:${1} --build-arg PYTHON_IMAGE=${1/-/:} -t apm-agent-python:${1} . # replace - with : to get the correct docker image
```
Comment (v1v) [inline tests/scripts/docker/run_tests.sh:50]: `elasticobservability/apm-agent-python-testing` ?

```suggestion
DOCKER_BUILDKIT=1 docker build --cache-from=elasticobservability/apm-agent-python-testing:${1} --build-arg PYTHON_IMAGE=${1/-/:} -t apm-agent-python:${1} . # replace - with : to get the correct docker image
```

**Comment by beniwohli** [comment]
Looking at a bunch of build logs, it looks like we're still spending a lot of time in docker `RUN` commands that should be cacheable, e.g. https://github.com/elastic/apm-agent-python/actions/runs/4617102660/jobs/8162941988?pr=1792#step:3:125. Am I misunderstanding something?

**Comment by reakaleek** [comment]
> Looking at a bunch of build logs, it looks like we're still spending a lot of time in docker `RUN` commands that should be cacheable, e.g. https://github.com/elastic/apm-agent-python/actions/runs/4617102660/jobs/8162941988?pr=1792#step:3:125. Am I misunderstanding something?

Yes. I still need to get following PRs merged to see the result.
- https://github.com/elastic/apm-agent-python/pull/1794
- https://github.com/elastic/apm-pipeline-library/pull/2174

**Review by beniwohli — APPROVED** [review]
Ah, gotcha. LGTM, then :)

**Comment by reakaleek** [comment]
In the latest run attempt it's still not utilising the cache as expected.

https://github.com/elastic/apm-agent-python/actions/runs/4617333116?pr=1792

But the images are definitely accessible from https://hub.docker.com/r/elasticobservability/apm-agent-python-testing/tags

Will need to debug a little further.

**Comment by reakaleek** [comment]
Okay, I give up. I tried multiple approaches and multiple combinations of using `--cache-from` without success.
Locally, it was working just fine. It seems like it just won't work in a GH Actions workflow.

Nevertheless, in it's current state the workflow run time decreased from ~10 minutes to ~4 minutes.
However, the ability to create changes in the docker set up for a specific commit is not given.

@basepi @beniwohli let me know what you think and if we can merge it for now.

If changing the docker setup frequently becomes a necessity, let's revisit it in the future.

WDYT?

**Review by basepi — APPROVED** [review]
Sounds fine to me. I'll re-request Beni's review and we can get this merged.

**Review by beniwohli — APPROVED** [review]
Really great work, thanks for taking this on, @reakaleek!

**Pull request merged**